qubesbuilder/plugins/source/scripts/get-and-verify-source

```diff
@@ -0,0 +1,392 @@
+        chmod 700 "$GNUPGHOME"
+        gpg --import "$KEYS_DIR/qubes-developers-keys.asc"
+        # Trust Qubes Master Signing Key
+        echo '427F11FD0FAA4B080123F01CDDFA1A3E36879494:6:' | gpg --import-ownertrust
+    fi
+    if [ "$KEYS_DIR/qubes-developers-keys.asc" -nt "$GNUPGHOME/pubring.gpg" ]; then
+        gpg --import "$KEYS_DIR/qubes-developers-keys.asc"
+        touch "$GNUPGHOME/pubring.gpg"
+    fi
+    for keyid in ${MAINTAINERS[*]}; do
+        gpg --import "$KEYS_DIR/$keyid.asc" || exit 1
+        echo "$keyid:6:" | gpg --import-ownertrust
+    done
+    gpgconf --kill gpg-agent
+
+    pushd "$REPO" >/dev/null || exit 2
+
+    expected_hash="$VERIFY_REF"
+    hash_len=${#expected_hash}
+    if [ "$hash_len" -ne 40 ] && [ "$hash_len" -ne 64 ]; then
+        echo "---> Bad Git hash value (wrong length); failing" >&2
+        exit 1
+    elif ! [[ "$expected_hash" =~ ^[a-f0-9]+$ ]]; then
+        echo "---> Bad Git hash value (bad character); failing" >&2
+        exit 1
+    fi
+
+    # Git format string, see man:git-for-each-ref(1) for details.
+    #
+    # The %(if)...%(then)...%(end) skips lightweight tags, which have no object to
+    # point to.  The colons allow a SHA-1 hash to be distinguished from a truncated
+    # SHA-256 hash, and also allow a truncated line to be detected.
+    format='%(if:equals=tag)%(objecttype)%(then)%(objectname):%(object):%(end)'
+    tags="$(git tag "--points-at=$expected_hash" "--format=$format" | head -c 500)" || exit 1
+    for tag in $tags; do
+        if ((${#tag} != hash_len * 2 + 2)); then
+            echo '---> Bad Git hash value (wrong length); failing' >&2
+            exit 1
+        elif ! [[ "${tag:hash_len}" == ":$expected_hash:" ]]; then
+            printf %s\\n "---> Tag has wrong hash (found ${tag:hash_len+1:hash_len}, expected $expected_hash)" >&2
+            exit 1
+        fi
+        tag="${tag:0:hash_len}"
+        if verify_git_obj tag "$tag"; then
+            echo "---> Good tag $tag"
+        else
+            echo "---> Invalid tag $tag"
+            exit 1
+        fi
+    done
+
+    if [ -z "${tag+x}" ]; then
+        echo "---> No tag pointing at $expected_hash"
+        if verify_git_obj commit "$expected_hash"; then
+            case $CHECK in
+            signed-tag-or-commit)
+                echo "---> $expected_hash does not have a signed tag"
+                echo "---> However, it is signed by a trusted key, and \$CHECK is set to $CHECK"
+                echo "---> Accepting it anyway"
+                ;;
+            signed-tag)
+                echo "---> $expected_hash is a commit signed by a trusted key ― did the signer forget to add a tag?"
+                exit 1
+                ;;
+            *)
+                echo "---> internal error (this is a bug)"
+                exit 1
+                ;;
+            esac
+        else
+            echo "---> Invalid commit $expected_hash"
+            exit 1
+        fi
```

Comment:
```suggestion
        if verify_git_obj commit "$expected_hash"; then
            case $CHECK in
            signed-tag-or-commit)
                echo "---> $expected_hash does not have a signed tag"
                echo "---> However, it is signed by a trusted key, and \$CHECK is set to $CHECK"
                echo "---> Accepting it anyway"
                ;;
            signed-tag)
                echo "---> $expected_hash is a commit signed by a trusted key ― did the signer forget to add a tag?"
                exit 1
                ;;
            *)
                echo "---> internal error (this is a bug)"
                exit 1
                ;;
            esac
        else
            echo "---> Invalid commit $expected_hash"
            exit 1
        fi >&2
```

Reply:
Added and this is in `fetch` plugin now.